### PR TITLE
Addon for operating system manager

### LINKED ISSUE #1719: Add support for operating system manager addon
**What feature would you like to be added?**
Support for operating system manager addon.

**What are use cases of the feature?**
[OSM](https://github.com/kubermatic/operating-system-manager) controller is responsible for generating bootstrapping(WIP) and provisioning configurations for **worker** machines. It decouples operating system configs into dedicated and isolable resources which are templatized; i.e. OperatingSystemProfiles. Thus providing a high degree of customizability for that configuration. 

When enabled, it works in liaison with [machine-controller](https://github.com/kubermatic/kubeone/tree/master/addons/machinecontroller) and takes over the responsibility of generating the OS configurations.

**Acceptance Criteria**
- Addon for OSM is added
- Webhooks for OSM are in-place
- machine-controller should switch to OSM for user data provisioning when OSM is enabled

### PR DESCRIPTION
**Pull request by ahmedwaleedmalik**
**What this PR does / why we need it**:
Introduce `operating-system-manager` as an optional addon.

- [x] Optional addon for OSM
- [x] Webhooks for OSM
- [x] Enable OSM in machine-controller if addon for OSM is enabled 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1719

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
* Addon for OperatingSystemManager
* machine-controller bumped to v1.42.2
```

### CODE REVIEW

File: pkg/tasks/tasks.go [hunk 1]
```diff
@@ -253,6 +255,18 @@ func WithResources(t Tasks) Tasks {
...
 				ErrMsg: "failed to wait for machine-controller",
 			},
+			{
+				Fn:          cloudinitsettings.Ensure,
+				ErrMsg:      "failed to create/configure cloud-init-settings namespace",
+				Description: "ensure cloud-init-settings namespace",
+				Predicate:   func(s *state.State) bool { return s.Cluster.AddonOperatingSystemManagerEnabled() },
+			},
+			{
+				Fn:          operatingsystemmanager.WaitReady,
```
Comment (kron4eg) [inline pkg/tasks/tasks.go:265]: I have gut feeling that we can combine this with machine-controller's. Just pass them some parameters as target for waiting.

Comment (ahmedwaleedmalik) [reply pkg/tasks/tasks.go:265]: We can but tbh I'd preferably keep this as a separate task. Eventually, OSM will be treated as a separate component/addon, we won't have to re-do stuff then.

**Comment by ahmedwaleedmalik** [comment]
This PR is currently blocked by https://github.com/kubermatic/operating-system-manager/pull/132

**Comment by ahmedwaleedmalik** [comment]
/hold

This PR requires a new release for [operating-system-manager](https://github.com/kubermatic/operating-system-manager)

File: addons/operating-system-manager/deployment-controller.yaml [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        "prometheus.io/path": "/metrics"
+      labels:
+        app: operating-system-manager
+    spec:
+      nodeSelector:
+        node-role.kubernetes.io/master: ""
+      tolerations:
+        - key: "node-role.kubernetes.io/master"
+          operator: Exists
+          effect: NoSchedule
```
Comment (xmudrii) [inline addons/operating-system-manager/deployment-controller.yaml:25]: We should also have `control-plane` toleration. You can check other addons as an example.

File: addons/operating-system-manager/deployment-controller.yaml [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+              path: /healthz
+              port: 8085
+            periodSeconds: 5
+          resources:
+            limits:
+              cpu: "2"
+              memory: 512Mi
+            requests:
+              cpu: 50m
+              memory: 128Mi
```
Comment (xmudrii) [inline addons/operating-system-manager/deployment-controller.yaml:84]: Please add a new line.

File: addons/operating-system-manager/webhook.yaml [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    metadata:
+      labels:
+        app: operating-system-manager-webhook
+    spec:
+      nodeSelector:
+        node-role.kubernetes.io/master: ""
+      tolerations:
+        - key: "node-role.kubernetes.io/master"
+          operator: Exists
+          effect: NoSchedule
```
Comment (xmudrii) [inline addons/operating-system-manager/webhook.yaml:48]: Please add the `control-plane` toleration.

File: pkg/addons/applier.go [hunk 5]
```diff
@@ -59,18 +60,20 @@ type applier struct {
...
-	Params                              map[string]string
+	Config                                   *kubeoneapi.KubeOneCluster
+	Certificates                             map[string]string
+	Credentials                              map[string]string
+	CredentialsCCM                           map[string]string
+	CCMClusterName                           string
+	CSIMigration                             bool
+	CSIMigrationFeatureGates                 string
+	MachineControllerCredentialsEnvVars      string
+	AddonOperatingSystemManagerEnabled       bool
```
Comment (xmudrii) [inline pkg/addons/applier.go:71]: I would drop this Addon prefix.

```suggestion
	OperatingSystemManagerEnabled       bool
```

File: pkg/addons/ensure.go [hunk 6]
```diff
@@ -35,27 +35,28 @@ var (
...
+		resources.AddonCNIWeavenet:            "",
+		resources.AddonCSIAwsEBS:              "",
+		resources.AddonCSIAzureDisk:           "",
+		resources.AddonCSIAzureFile:           "",
+		resources.AddonCSIDigitalOcean:        "",
+		resources.AddonCSIHetzner:             "",
+		resources.AddonCSIOpenStackCinder:     "",
+		resources.AddonCSIVsphere:             "",
+		resources.AddonMachineController:      "",
+		resources.AddonOperatingSystemManager: "",
```
Comment (xmudrii) [inline pkg/addons/ensure.go:57]: I don't think we need to add this here. Let's treat OSM as an "accessible" addon for the beginning, not an embedded addon.

File: pkg/apis/kubeone/helpers.go [hunk 7]
```diff
@@ -88,6 +90,17 @@ func (h *HostConfig) SetLeader(leader bool) {
 	h.IsLeader = leader
 }
 
+func (c KubeOneCluster) AddonOperatingSystemManagerEnabled() bool {
```
Comment (xmudrii) [inline pkg/apis/kubeone/helpers.go:93]: Same as the comment related to the variable name, I'd drop this `Addon` prefix.

```suggestion
func (c KubeOneCluster) OperatingSystemManagerEnabled() bool {
```

File: pkg/cloudinitsettings/resources.go [hunk 8]
```diff
@@ -0,0 +1,117 @@
+/*
```
Comment (xmudrii) [inline pkg/cloudinitsettings/resources.go:1]: Why is this not an addon?

File: pkg/credentials/secret.go [hunk 9]
```diff
@@ -75,14 +77,21 @@ func Ensure(s *state.State) error {
...
-		return errors.Wrap(createErr, "failed to ensure credentials secret")
+		return errors.Wrap(createErr, "failed to ensure credentials secret for machine-controller")
+	}
+
+	if s.Cluster.AddonOperatingSystemManagerEnabled() {
+		osmSecret := credentialsSecret(SecretNameOSM, providerCreds)
+		if createErr := clientutil.CreateOrReplace(context.Background(), s.DynamicClient, osmSecret); createErr != nil {
+			return errors.Wrap(createErr, "failed to ensure credentials secret for operating-system-manager")
+		}
 	}
```
Comment (xmudrii) [inline pkg/credentials/secret.go:95]: Why do we need a dedicated secret for OSM? Can we just reuse the MC secret?

File: pkg/addons/ensure.go [hunk 10]
```diff
@@ -106,6 +107,12 @@ func collectAddons(s *state.State) (addonsToDeploy []addonAction) {
 		})
 	}
 
+	if s.Cluster.AddonOperatingSystemManagerEnabled() {
+		addonsToDeploy = append(addonsToDeploy, addonAction{
+			name: resources.AddonOperatingSystemManager,
+		})
+	}
```
Comment (xmudrii) [inline pkg/addons/ensure.go:114]: I think we can just deploy OSM as an accessible/user addon.

File: pkg/cloudinitsettings/resources.go (hunk 8 above)
Comment (ahmedwaleedmalik) [reply pkg/cloudinitsettings/resources.go:1]: I'm not sure if this qualifies as an addon, this is more like a pre-requisite for machine-controller to work with OSM.

File: pkg/credentials/secret.go (hunk 9 above)
Comment (ahmedwaleedmalik) [reply pkg/credentials/secret.go:95]: We can but I'm not really a big fan of sharing secrets between distinct apps, even if they are exactly the same. But yes we can probably do that in this case, should I?

File: pkg/cloudinitsettings/resources.go (hunk 8 above)
Comment (ahmedwaleedmalik) [reply pkg/cloudinitsettings/resources.go]: Done

**Review by kron4eg — APPROVED** [review]
/approve

**Review by kron4eg — APPROVED** [review]

**Pull request merged**